src/static-build/loader.ts

```diff
@@ -0,0 +1,150 @@
+import getFeatures from './getFeatures';
+import { LoaderContext } from 'webpack/lib/webpack';
+const { getOptions } = require('loader-utils');
+import * as recast from 'recast';
+const types = recast.types;
+const namedTypes = types.namedTypes;
+const builders = types.builders;
+const compose = require('recast/lib/util').composeSourceMaps;
+
+/**
+ * A map of features that should be statically replaced in the code
+ */
+export interface StaticHasFeatures {
+	[feature: string]: boolean;
+}
+
+const HAS_MID = /\/has$/;
+const HAS_PRAGMA = /^\s*(!?)\s*has\s*\(["']([^'"]+)['"]\)\s*$/;
+
+export default function (this: LoaderContext, content: string, sourceMap?: { file: '' }) {
+	// copy features to a local scope, because `this` gets weird
+	const options = getOptions(this);
+	const { features: featuresOption, isRunningInNode } = options;
+	const args = (sourceMap && sourceMap.file && {
+		sourceFileName: sourceMap.file
+	}) || undefined;
+	let features: StaticHasFeatures;
+
+	if (!featuresOption || Array.isArray(featuresOption) || typeof featuresOption === 'string') {
+		features = getFeatures(featuresOption, isRunningInNode);
+	}
+	else {
+		features = featuresOption;
+	}
+	const dynamicFlags = new Set<string>();
+	const ast = recast.parse(content, args);
```

Comment:
do you think we should only parse and traverse after a basic string sniff? this is potentially pretty costly to run on every file.

Reply:
Good idea. I added a check for any has pragmas or calls to `require(*/has)` before parsing.